Remarks from review #233 (#240)
codewars/docs: content/authoring/guidelines/submission-tests.md

## General Guidelines

- **Conform to [General Coding Guidelines][authoring-guidelines-general-coding]**: tests are code too and as such should keep up to code quality standards.

> - **Tests should verify outcomes of solutions, and not their specific approach or implementation details.** A solution which returns correct answers and satisfies all required charactristics should be accepted regardless of used algorithm or details of implementation.

Reviewer: ```suggestion
- **Tests should verify outcomes of solutions, and not their specific approach or implementation details.** A solution that returns correct answers and satisfies all required characteristics should be accepted regardless of the used algorithm or details of implementation.
```

Some kata require solutions to be fast enough. For example, the author may only wish to accept solutions completing in (sub-)linear time. Building such test suites is not an easy task!

- **Performance tests can be implemented in terms of random tests, by testing with large random inputs.** However, to make debugging easier, it can be worth having a separate set of random tests with small inputs first.

> - If possible, **pefrormance of a solution should be evaluated by gauging the size of the inputs, rather than amount of tests.** If the difficulty of a kata is roughly proportional to the size of the input, it's usually better to have a few tests with large input rather than many with medium-sized inputs. For example, 100 tests with huge numbers or arrays is usually better than 1000 tests with moderately large arrays/numbers. At the same time, authors should ensure that large inputs are generated and handled in a way which does not have too much negative impact on the test suite.

Author: ```suggestion
- If possible, **performance of a solution should be evaluated by gauging the size of the inputs, rather than amount of tests.** If the difficulty of a kata is roughly proportional to the size of the input, it's usually better to have a few tests with large input rather than many with medium-sized inputs. For example, 100 tests with huge numbers or arrays is usually better than 1000 tests with moderately large arrays/numbers. At the same time, authors should ensure that large inputs are generated and handled in a way which does not have too much negative impact on the test suite.
```

> - If possible, **performance of a solution should be evaluated by gauging the size of the inputs, rather than amount of tests.** If the difficulty of a kata is roughly proportional to the size of the input, it's usually better to have a few tests with large input rather than many with medium-sized inputs. For example, 100 tests with huge numbers or arrays is usually better than 1000 tests with moderately large arrays/numbers. At the same time, authors should ensure that large inputs are generated and handled in a way which does not have too much negative impact on the test suite.

Reviewer: ```suggestion
- If possible, **performance of a solution should be evaluated by gauging the size of the inputs, rather than the amount of tests.** If the difficulty of a kata is roughly proportional to the size of the input, it's usually better to have a few tests with large input rather than many with medium-sized inputs. For example, 100 tests with huge numbers or arrays is usually better than 1000 tests with moderately large arrays/numbers. At the same time, authors should ensure that large inputs are generated and handled in a way that does not have too much negative impact on the test suite.
```

- **Performance tests should be consistent between runs.** It should not happen that one and the same solution sometimes passes, and sometimes fails, depending on randomized inputs.
- **Make sure that what you measure is what you want**, when solutions are to be rejected based on their performance. With huge inputs, the random generation of the inputs may be more time-consuming than the computation of the expected result. Hence the overall timing indication in the output panel is generally useless to ensure that the performance tests are actually discriminating the different kinds of solutions as expected. The evaluation of the time actually used by the user's solution should be done (and compared to the reference solution if any) excluding the input generation time.

> - **Kata should not call for micro-optimizations where not necessary.** Performance tests should leave some freedom for users and give some leeway to solutions based on various approaches to the problem or with slight differences in implementation, as long as they satisfy general performance criteria.

Reviewer: ```suggestion
- **Kata should not call for micro-optimizations when not necessary.** Performance tests should leave some freedom for users and give some leeway to solutions based on various approaches to the problem or with slight differences in implementation, as long as they satisfy general performance criteria.
```
